### PR TITLE
feat: ETH API: Add V1 API to convert Filecoin Address to ETH address

### PR DESCRIPTION
**Pull request by aarshkshah1992**
Closes https://github.com/filecoin-project/lotus/issues/12308.

Introduces a new `FilecoinAddressToEthAddressV1` API that allows all Filecoin address types (f0/1/2/3/4) to ETH addresses based on the user's reorg tolerance.

Deprecates the existing `FilecoinAddressToEthAddress` API that only supports `f0`/`f4` addresses in favour of `FilecoinAddressToEthAddressV1`.

### CODE REVIEW

File: node/impl/full/eth.go [hunk 1]
```diff
@@ -207,6 +208,56 @@ func (a *EthAPI) FilecoinAddressToEthAddress(ctx context.Context, filecoinAddres
...
+	case address.SECP256K1, address.Actor, address.BLS:
+		// Valid protocols
+	default:
+		// Ideally, this should never happen but is here for sanity checking.
+		return ethtypes.EthAddress{}, xerrors.Errorf("invalid filecoin address protocol: %s", filecoinAddress.String())
+	}
+
+	// Set default block parameter to "finalized" if not provided
+	if blkParam == "" || len(blkParam) == 0 {
+		blkParam = "finalized"
```
Comment (aarshkshah1992) [inline node/impl/full/eth.go:230]: @rvagg Going with "finalized" as the default here as we want the default to be the safest option here. 

The other thing we can do here is to get rid of the default and fail the request if client hasn't specified a `blkParam` for a f1/f2/f3 address.

File: itests/eth_api_test.go [hunk 2]
```diff
@@ -92,6 +98,122 @@ func TestFilecoinAddressToEthAddress(t *testing.T) {
...
+
+	idAddr, err = client.StateLookupID(ctx, multisigAddress, types.EmptyTSK)
+	require.NoError(t, err)
+
+	ethAddr, err = ethtypes.EthAddressFromFilecoinAddress(idAddr)
+	require.NoError(t, err)
+	require.Equal(t, ethAddr, apiEthAddr)
+}
+
+func TestFilecoinAddressToEthAddressV1Finalised(t *testing.T) {
```
Comment (aarshkshah1992) [inline itests/eth_api_test.go:181]: @rvagg This test tests for the "finalized" blkparam for an f1 address.

File: node/impl/full/eth.go (hunk 1 above)
Comment (rvagg) [reply node/impl/full/eth.go:230]: finalized seems fine to me, and _should_ be awesome post-f3.

File: node/impl/full/eth.go [hunk 3]
```diff
@@ -207,6 +208,56 @@ func (a *EthAPI) FilecoinAddressToEthAddress(ctx context.Context, filecoinAddres
...
+	ts, err := getTipsetByBlockNumber(ctx, a.Chain, blkParam, true)
+	if err != nil {
+		return ethtypes.EthAddress{}, xerrors.Errorf("failed to get tipset for block %s: %w", blkParam, err)
+	}
+
+	// Lookup the ID address
+	idAddr, err := a.StateManager.LookupIDAddress(ctx, filecoinAddress, ts)
+	if err != nil {
+		return ethtypes.EthAddress{}, xerrors.Errorf(
+			"failed to lookup ID(f0) address for given Filecoin address %s: "+
```
Comment (rvagg) [inline node/impl/full/eth.go:243]: it's a `t0` on calibnet, although I'm not sure of the best logic to swap that out here dynamically but I think you can call out to `address.CurrentNetwork`, check whether it's `address.Testnet` or `address.Mainnet` then use either `address.TestnetPrefix` or `address.MainnetPrefix` (although I bet the best way is to do something like `address.NewIDAddress(0)`, `.String()` it and cut off the first 2 chars).

**Comment by rvagg** [comment]
Implementation seems fine to me, but we can't call this `FilecoinAddressToEthAddressV1`, that's not going to work. For a start it's the second iteration of this API, not the first, but also this is on the /v1/ RPC API, so that's just weird. Making it `V2` hanging off a /v1/ is going to be extra weird.

We have a couple of problems to wrestle with:

1. We probably should stop evolving the /v1/ API as @jennijuju mentioned on Slack, but I'm not sure that helps us here because do we really want to make the start of a /v2/ a prerequisite for getting this done?
2. We don't want to break any /v1/ APIs
3. We can't add optional arguments to these calls, it's not how the RPC system (or at least our go-jsonrpc) works

I proposed two alternative solutions:

* Use the name `EthAddressFromFilecoinAddress`, which mirrors the existing `EthAddressToFilecoinAddress`, perhaps we then drop `FilecoinAddressToEthAddress` from the /v2/ API. Unfortunately this would be adding another method to /v1/ but I don't see a way around that atm.
* Make the existing `FilecoinAddressToEthAddress` handle this, but do it in a backward compatible way by hijacking the JSON encoding, which I proposed in: https://github.com/filecoin-project/lotus/pull/12323

I'd like to hear @Stebalien's thoughts on this though.

**Comment by rvagg** [comment]
@aarshkshah1992 I just stumbled upon this, after tinkering on the gateway: https://github.com/filecoin-project/ref-fvm/issues/1573

For `eth_subscribe` the API has an optional parameter: when you say `"newHeads"` it's just 1 param, but when you say `"logs"` you supply a second optional parameter: https://docs.infura.io/api/networks/ethereum/json-rpc-methods/subscription-methods/eth_subscribe

It was implemented in https://github.com/filecoin-project/lotus/pull/10027.

That PR has a lot of noise to support the subscribe data flow so it needs a whole reverse implementation for the client. But the relevant bit is the replacement of the method's params with `params jsonrpc.RawParams`. If the method signature is `(ctx context.Context, params jsonrpc.RawParams)` then go-jsonrpc will just give you the `json.RawMessage` to decode for yourself.

There are 3 methods that do this in the codebase so far:
* `EthFeeHistory`: https://github.com/filecoin-project/lotus/blob/f6978f01725fc8f8ef72cdf83d15aa57b8e076db/node/impl/full/eth.go#L695
* `EthEstimateGas`: https://github.com/filecoin-project/lotus/blob/f6978f01725fc8f8ef72cdf83d15aa57b8e076db/node/impl/full/eth.go#L1244
* `EthSubscribe`: https://github.com/filecoin-project/lotus/blob/f6978f01725fc8f8ef72cdf83d15aa57b8e076db/node/impl/full/eth.go#L1875

All with the same pattern.

So, I think you can just do this for `FilecoinAddressToEthAddress` and add your optional parameter on the end and we have a non-breaking change.

**Comment by aarshkshah1992** [comment]
@rvagg 

Also tested this on a devnet node

```
curl -X POST -H "Content-Type: application/json" --data '{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "Filecoin.FilecoinAddressToEthAddress",
  "params": ["t410fghja63ghyzd54cbswfapglsutx5e64xnm7mazwi"]
}' http://localhost:1234/rpc/v1
{"id":1,"jsonrpc":"2.0","result":"0x31d20f6cc7c647de0832b140f32e549dfa4f72ed"}
```

```
curl -X POST -H "Content-Type: application/json" --data '{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "Filecoin.FilecoinAddressToEthAddress",
  "params": ["t3thonru4djgcwi2vuuymzqkbymiv6tixs5igehh3ht46tjgqyjifkuz2sbziovsgshr2vttjh34q6bzavelea","latest"]
}' http://localhost:1234/rpc/v1
{"id":1,"jsonrpc":"2.0","result":"0xff00000000000000000000000000000000000064"}
```

File: api/api_full.go [hunk 4]
```diff
@@ -726,8 +726,33 @@ type FullNode interface {
...
+	// "f0" ID addresses are not permanent and can be affected by chain reorganizations. To account for this,
+	// the API includes a `blkNum` parameter, which specifies the block number that is used to determine the tipset state to use for converting an
+	// "f1"/"f2"/"f3" address to an "f0" address. This parameter functions similarly to the `blkNum` parameter in the existing `EthGetBlockByNumber` API.
+	// See https://docs.alchemy.com/reference/eth-getblockbynumber for more details.
+	//
+	// Parameters:
+	// - ctx: The context for the API call.
+	// - filecoinAddress: The Filecoin address to convert.
+	// - blkNum: The block number or state for the conversion. Defaults to "finalized" for maximum safety.
+	//   Possible values: "pending", "latest", "finalized", "safe", or a specific block number.
```
Comment (rvagg) [inline api/api_full.go:749]:
```suggestion
	//   Possible values: "pending", "latest", "finalized", "safe", or a specific block number represented as hex.
```

File: api/api_full.go [hunk 5]
```diff
@@ -726,8 +726,33 @@ type FullNode interface {
 	EthAccounts(ctx context.Context) ([]ethtypes.EthAddress, error) //perm:read
 	// EthAddressToFilecoinAddress converts an EthAddress into an f410 Filecoin Address
 	EthAddressToFilecoinAddress(ctx context.Context, ethAddress ethtypes.EthAddress) (address.Address, error) //perm:read
-	// FilecoinAddressToEthAddress converts an f410 or f0 Filecoin Address to an EthAddress
-	FilecoinAddressToEthAddress(ctx context.Context, filecoinAddress address.Address) (ethtypes.EthAddress, error) //perm:read
+
+	// `FilecoinAddressToEthAddress` is the recommended API for converting any Filecoin address to an EthAddress.
```
Comment (rvagg) [inline api/api_full.go:730]:
```suggestion
	// `FilecoinAddressToEthAddress` converts any Filecoin address to an EthAddress.
```

File: api/api_full.go [hunk 6]
```diff
@@ -726,8 +726,33 @@ type FullNode interface {
...
+	//
+	// This method supports all Filecoin address types:
+	// - "f0" and "f4" addresses: Converted directly.
+	// - "f1", "f2", and "f3" addresses: First converted to their corresponding "f0" ID address, then to an EthAddress.
+	//
+	// Requirements:
+	// - For "f1", "f2", and "f3" addresses, they must be instantiated on-chain, as "f0" ID addresses are only assigned to actors when they are created on-chain.
```
Comment (rvagg) [inline api/api_full.go:737]: I wonder if we should go into detail about this? I remember being confused about this the first time I encountered it; it's perfectly logical to me now but for the uninitiated they may need to know a bit more about what it means to be on chain

File: build/openrpc/full.json [hunk 7]
```diff
@@ -6922,21 +6922,29 @@
...
+                                "description": "Number is a number",
+                                "type": [
+                                    "number"
+                                ]
+                            }
                         ],
-                        "additionalProperties": false,
                         "type": [
-                            "object"
+                            "array"
```
Comment (rvagg) [inline build/openrpc/full.json:6947]: we're going to need to do something about these specs for correctness

File: chain/types/ethtypes/eth_types.go [hunk 8]
```diff
@@ -826,6 +826,44 @@ func GetContractEthAddressFromCode(sender EthAddress, salt [32]byte, initcode []
 	return ethAddr, nil
 }
 
+type FilecoinAddressToEthAddressParams struct {
+	FilecoinAddress address.Address
+	BlkParam        *string
```
Comment (rvagg) [inline chain/types/ethtypes/eth_types.go:831]:
```suggestion
	FilecoinAddress address.Address `json:"address"`
	Block        *string `json:"block,omitempty"`
```

Are we following a pattern with `BlkParam`? seems odd to name it

this `json` tagging isn't really critical since this is internal, just nice to be consistent

Comment (rvagg) [reply chain/types/ethtypes/eth_types.go:831]: actually forget this, it's never touched

File: chain/types/ethtypes/eth_types.go [hunk 9]
```diff
@@ -826,6 +826,44 @@ func GetContractEthAddressFromCode(sender EthAddress, salt [32]byte, initcode []
...
+		}
+	default:
+		return xerrors.Errorf("expected 1 or 2 params, got %d", len(params))
+	}
+
+	return nil
+}
+
+func (e *FilecoinAddressToEthAddressParams) MarshalJSON() ([]byte, error) {
+	if e.BlkParam != nil {
```
Comment (rvagg) [inline chain/types/ethtypes/eth_types.go:861]:
```suggestion
	if e.BlkParam != nil && e.BlkParam != "" {
```

File: itests/eth_api_test.go [hunk 10]
```diff
@@ -52,44 +60,147 @@ func TestEthAddressToFilecoinAddress(t *testing.T) {
...
-	require.NoError(t, err)
+		return jsonrpc.RawParams(params)
+	}
 
+	// Test for f4 address
+	filecoinDelegatedAddr, err := client.WalletNew(ctx, types.KTDelegated)
+	require.NoError(t, err)
+	ethAddr, err := ethtypes.EthAddressFromFilecoinAddress(filecoinDelegatedAddr)
+	require.NoError(t, err)
+	apiEthAddr, err := client.FilecoinAddressToEthAddress(ctx, mkParams(filecoinDelegatedAddr, ""))
```
Comment (rvagg) [inline itests/eth_api_test.go:89]: if you look at the other uses of `RawParams` in itests there's a use of lib/result/result.Wrap that makes this a bit easier and avoids the extra helper function:

```go
	history, err := client.EthFeeHistory(ctx, result.Wrap[jsonrpc.RawParams](
		json.Marshal([]interface{}{5, "0x10"}),
	).Assert(require.NoError))
```

File: node/impl/full/eth.go [hunk 11]
```diff
@@ -203,8 +204,63 @@ func (a *EthAPI) EthAddressToFilecoinAddress(ctx context.Context, ethAddress eth
...
+		return ethtypes.EthAddress{}, xerrors.Errorf("failed to get tipset for block %s: %w", blkParam, err)
+	}
+
+	// Lookup the ID address
+	idAddr, err := a.StateManager.LookupIDAddress(ctx, filecoinAddress, ts)
+	if err != nil {
+		return ethtypes.EthAddress{}, xerrors.Errorf(
+			"failed to lookup ID address for given Filecoin address %s: "+
+				"ensure that the address has been instantiated on-chain and sufficient epochs have passed since instantiation to confirm to the given 'blkParam' %s. "+
+				"Error: %w",
```
Comment (rvagg) [inline node/impl/full/eth.go:250]:
```suggestion
			"failed to lookup ID address for given Filecoin address %s ("+
				"ensure that the address has been instantiated on-chain and sufficient epochs have passed since instantiation to confirm to the given 'blkParam': \"%s\"): %w",
```

Mainly making this not require a full-stop and capital `"Error"`.

Although, I also think you should pull that entire bracket part out and make it conditional, and you could have 3 cases for it:

1. they have a `blkParam` that's not hex, so is one of the text ones, in which case this message is 👌 
2. they have a `blkParam` but it's a hex, in which case you could have an alternate message about how it's not found at that block
3. they have no `blkParam`, in which case they get told to wait for `"finalized"` but they never said such a thing, you could reword it to inform about the default

File: itests/eth_api_test.go (hunk 10 above)
Comment (aarshkshah1992) [reply itests/eth_api_test.go]: Done.

**Comment by aarshkshah1992** [comment]
@rvagg Addressed your review. Please can you 🟢  ?

File: CHANGELOG.md [hunk 12]
```diff
@@ -19,6 +19,9 @@
 
 - feat: Add trace filter API supporting RPC method `trace_filter` ([filecoin-project/lotus#12123](https://github.com/filecoin-project/lotus/pull/12123)). Configuring `EthTraceFilterMaxResults` sets a limit on how many results are returned in any individual `trace_filter` RPC API call.
 
+- feat: `FilecoinAddressToEthAddress` RPC can now return ETH addresses for all Filecoin address types ("f0"/"f1"/"f2"/"f3") based on client's re-org tolerance.
```
Comment (rvagg) [inline CHANGELOG.md:22]:
```suggestion
- feat: `FilecoinAddressToEthAddress` RPC can now return ETH addresses for all Filecoin address types ("f0"/"f1"/"f2"/"f3") based on client's re-org tolerance. This is a breaking change if you are using the API via the go-jsonrpc library or by using Lotus as a library, but is a non-breaking change when using the API via any other RPC method as it adds an optional second argument.
```

**Review by rvagg — APPROVED** [review]
I'm a tiny bit hesitant because it's quite disruptive when you see how you have to deal with the API in Go via Lotus directly, but I'm going to make the assumption that the majority of callers of this method are doing it via plain RPC or some other RPC wrapper.

**Comment by aarshkshah1992** [comment]
@rvagg IIUC, most clients do indeed consume this via RPC so I think the cost is worth it as it allows to get away without creating a new API and confusing RPC users.
